File: AWS/iac-microbenchmark/ec2/rust/aes256_encrypt/src/main.rs
use aws_sdk_kms::{self as kms, primitives::Blob};
use base64::encode;
use serde::{Deserialize, Serialize};
use std::env;
use std::error::Error;

#[derive(Serialize, Deserialize, Debug)]
struct MessageStruct {
    message: String,
}

#[derive(Serialize, Deserialize, Debug)]
struct EncryptedMessageStruct {
    ciphertext: String,
    iv: String,
    tag: String,
    encrypted_key: String,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // Get the KMS key ARN from environment variables
    let aes_kms_key_id = env::var("AES_KMS_KEY_ARN").map_err(|_| "KMS key ARN not set")?;

    // Get JSON input from command line
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        return Err("Usage: ./rustexecutable '{\"message\":\"your_message_here\"}'".into());
    }

    let body_str = &args[1];

    // Parse the input JSON
    let message_struct: MessageStruct = serde_json::from_str(body_str)
        .map_err(|_| "Invalid Request Body, missing message or incorrect format.")?;

    // Convert the message to bytes
    let message_blob = Blob::new(message_struct.message.as_bytes());

    // Initialize the KMS client
    let config = aws_config::load_from_env().await;
    let kms_client = kms::Client::new(&config);

    // Attempt to encrypt the message using KMS
    match kms_encrypt_message(&kms_client, &aes_kms_key_id, message_blob).await {
        Ok(encrypted_message_struct) => {
            // Serialize the struct to a JSON string
            let str_encrypted_message_struct = serde_json::to_string(&encrypted_message_struct)?;
            println!("{}", str_encrypted_message_struct);
            Ok(())
        }
        Err(e) => Err(format!("Error encrypting message: {}", e).into()),
    }
}

/// Function to encrypt a message using AWS KMS. AES 256
async fn kms_encrypt_message(
    kms_client: &kms::Client,
    key_id: &str,
    message: Blob,
) -> Result<EncryptedMessageStruct, Box<dyn Error>> {
    // Generate a data key using KMS
    let data_key_output = kms_client
        .generate_data_key()
        .key_id(key_id)
        .key_spec("AES_256".into())
        .send()
        .await
        .map_err(|err| format!("Failed to generate data key: {}", err))?;

    let plaintext_data_key = data_key_output
        .plaintext
        .ok_or_else(|| "No plaintext data key returned")?;

    let plaintext_dk_u8 = plaintext_data_key.into_inner(); // Convert

    let encrypted_data_key = data_key_output
        .ciphertext_blob
        .ok_or_else(|| "No encrypted data key returned")?;

    // Generate IV and tag for AES-GCM
    let mut iv = [0u8; 16]; // Initialize a 16-byte array for IV
    openssl::rand::rand_bytes(&mut iv).map_err(|_| "Failed to generate IV")?;

    // AES-GCM encryption using OpenSSL
    let cipher = openssl::symm::Cipher::aes_256_gcm();
    let mut tag = vec![0; 16]; // 16-byte tag for AES-GCM

    let ciphertext = openssl::symm::encrypt_aead(
        cipher,
        &plaintext_dk_u8,
        Some(&iv),        // Use the IV we generated
        &[],              // No additional authenticated data (AAD)
        message.as_ref(), // The message to encrypt
        &mut tag,         // Output tag buffer
    )
    .map_err(|_| "AES Encryption failed")?;

    // Create the encrypted message struct, with Base64-encoded values
    let encrypted_message_struct = EncryptedMessageStruct {
        ciphertext: encode(&ciphertext),
        iv: encode(&iv),
        tag: encode(&tag),
        encrypted_key: encode(&encrypted_data_key),
    };

    Ok(encrypted_message_struct)
}

File: AWS/iac-microbenchmark/ec2/go/aes256_encrypt/main.go
// main.go
package main

import (
	"context"
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"os"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/kms"
	"github.com/aws/aws-sdk-go-v2/service/kms/types"
)

// Struct to parse the input JSON message
type MessageStruct struct {
	Message string `json:"message"`
}

// Struct for the encrypted response output
type EncryptedResponse struct {
	Ciphertext   string `json:"ciphertext"`
	EncryptedAesKey string `json:"encrypted_aes_key"`
	IV           string `json:"iv"`
	Tag          string `json:"tag"`
	Error        string `json:"error,omitempty"`
}

func main() {
	// Read JSON input from command-line argument
	requestJSONRaw := os.Args[1]
	var messageStruct MessageStruct

	// Parse the JSON input
	if err := json.Unmarshal([]byte(requestJSONRaw), &messageStruct); err != nil || messageStruct.Message == "" {
		fmt.Println(`{"error": "Invalid input: missing or malformed message"}`)
		return
	}

	// Retrieve the KMS key ARN from the environment
	kmsKeyID := os.Getenv("AES_KMS_KEY_ARN")
	if kmsKeyID == "" {
		fmt.Println(`{"error": "KMS key ARN environment variable not set"}`)
		return
	}

	context := context.Background()
	// Load AWS configuration
	cfg, err := config.LoadDefaultConfig(context)
	if err != nil {
		fmt.Println(`{"error": "Failed to load AWS config"}`)
		return
	}

	// Create a KMS client
	kmsClient := kms.NewFromConfig(cfg)

	// Generate a data key using KMS
	dataKeyOutput, err := kmsClient.GenerateDataKey(context, &kms.GenerateDataKeyInput{
		KeyId:   aws.String(kmsKeyID),
		KeySpec: types.DataKeySpecAes256,
	})
	if err != nil {
		fmt.Printf(`{"error": "Failed to generate data key: %v"}`, err)
		return
	}

	// Extract plaintext and encrypted data key
	plaintextDataKey := dataKeyOutput.Plaintext
	encryptedDataKey := dataKeyOutput.CiphertextBlob

	// Encrypt the message using AES-GCM
	ciphertext, iv, tag, err := encryptMessageWithAESGCM(plaintextDataKey, []byte(messageStruct.Message))
	if err != nil {
		fmt.Printf(`{"error": "Failed to encrypt message: %v"}`, err)
		return
	}

	// Prepare the response
	response := EncryptedResponse{
		Ciphertext:   base64.StdEncoding.EncodeToString(ciphertext),
		EncryptedAesKey: base64.StdEncoding.EncodeToString(encryptedDataKey),
		IV:           base64.StdEncoding.EncodeToString(iv),
		Tag:          base64.StdEncoding.EncodeToString(tag),
	}
	responseJSON, _ := json.Marshal(response)
	fmt.Println(string(responseJSON))
}

// AES-GCM encryption function
func encryptMessageWithAESGCM(key []byte, message []byte) ([]byte, []byte, []byte, error) {
	ivSize := 12

	// Create a new AES cipher block
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, nil, nil, fmt.Errorf("failed to create AES cipher: %w", err)
	}

	// Use GCM mode for encryption with a specified IV size
	gcm, err := cipher.NewGCMWithNonceSize(block, ivSize)
	if err != nil {
		return nil, nil, nil, fmt.Errorf("failed to create GCM cipher: %w", err)
	}

	// Generate a random IV
	iv := make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		return nil, nil, nil, fmt.Errorf("failed to generate IV: %w", err)
	}

	// Encrypt the message using AES-GCM
	ciphertext := gcm.Seal(nil, iv, message, nil)
	tag := ciphertext[len(ciphertext)-gcm.Overhead():]
	ciphertext = ciphertext[:len(ciphertext)-gcm.Overhead()]

	return ciphertext, iv, tag, nil
}
